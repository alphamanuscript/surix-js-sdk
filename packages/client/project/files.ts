import { AxiosInstance } from 'axios';
import { ApiFile, ProjectFiles } from '../types';

export function getProjectFiles (projectId: string, apiClient: AxiosInstance): ProjectFiles {
  return {
    async get (fileId: string): Promise<ApiFile> {
      const res = await apiClient.get<ApiFile>(`/projects/${projectId}/files/${fileId}`);
      return res.data;
    },
    async list (): Promise<ApiFile[]> {
      const res = await apiClient.get<ApiFile[]>(`/projects/${projectId}/files`);
      return res.data;
    }
  };
}