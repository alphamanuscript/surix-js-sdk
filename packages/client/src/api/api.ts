import axios, { AxiosInstance } from 'axios';

export function getApiClient (baseUrl: string, apiKey: string): AxiosInstance {
  const axiosInstance = axios.create({
    baseURL: baseUrl
  })
  axiosInstance.defaults.headers['Authorization'] = `apikey ${apiKey}`
  return axiosInstance
}