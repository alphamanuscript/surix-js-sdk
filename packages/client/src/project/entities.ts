import { 
  ApiEntity,  
  expandEntity, 
  expandQuery, 
  wrapEntity, 
  wrapEntityArray, 
  WrappedEntity
} from '@surix/data-helpers';
import { AxiosInstance } from 'axios';
import {  DeletedEntities, Entity, ProjectEntities, Query } from '../types';

export function getProjectEntities (projectId: string, apiClient: AxiosInstance): ProjectEntities {
  return {
    async get(entityId: string): Promise<WrappedEntity> {
      const res = await apiClient.get<ApiEntity>(`/projects/${projectId}/entities/${entityId}`);
      return wrapEntity(res.data);
    },
    async query(query: Query = {}): Promise<WrappedEntity[]> {
      const expandedQuery = expandQuery(query);
      if (Object.keys(expandedQuery.query).length === 0) {
        delete expandedQuery.query;
      }
      const res = await apiClient.post<ApiEntity[]>(
        `/projects/${projectId}/entities/query`, expandedQuery);
      const entities = res.data;
      return wrapEntityArray(entities);
    },
    async update(entity: Entity): Promise<WrappedEntity> {
      const _entity = { ...entity };
      const entityId = _entity._id;
      delete _entity._id;
      const expandedEntity = expandEntity(_entity);
      const res = await apiClient.patch<ApiEntity>(
        `/projects/${projectId}/entities/${entityId}`, expandedEntity);
      const returnedEntity = res.data;
      return wrapEntity(returnedEntity);
    },
    async put(entity: Entity): Promise<WrappedEntity> {
      const _entity = { ...entity }; // Make a shallow copy
      const entityId = _entity._id;
      delete _entity._id;
      const expandedEntity = expandEntity(_entity);
      const res = await apiClient.put<ApiEntity>(
        `/projects/${projectId}/entities/${entityId}`, expandedEntity);
      const returnedEntity = res.data;
      return wrapEntity(returnedEntity);
    },
    async delete(entityId: string): Promise<WrappedEntity> {
      const res = await apiClient.delete(`/projects/${projectId}/entities/${entityId}`);
      const returnedEntity = res.data;
      return wrapEntity(returnedEntity);
    },
    async deleteMany(entityIds: string[]): Promise<DeletedEntities> {
      const res = await apiClient.delete(`/projects/${projectId}/entities`, 
      {
        data: {
          entities: entityIds
        }
      });
      return res.data;
    },
    async create(entity: Entity): Promise<WrappedEntity> {
      const expandedEntity = expandEntity(entity);
      const res = await apiClient.post<ApiEntity>(
        `/projects/${projectId}/entities`, expandedEntity);
      const returnedEntity = res.data;
      return wrapEntity(returnedEntity);
    }
  };
}