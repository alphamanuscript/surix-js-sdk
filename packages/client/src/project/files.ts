import {
  ApiFile,
  wrapFile,
  wrapFileArray,
  WrappedFile } from '@surix/data-helpers';
import { AxiosInstance } from 'axios';
import { ProjectFiles} from '../types';

export function getProjectFiles (projectId: string, apiClient: AxiosInstance): ProjectFiles {
  return {
    async get (fileId: string): Promise<WrappedFile> {
      const res = await apiClient.get<ApiFile>(`/projects/${projectId}/files/${fileId}`);
      return wrapFile(res.data);
    },
    async list (): Promise<WrappedFile[]> {
      const res = await apiClient.get<ApiFile[]>(`/projects/${projectId}/files`);
      return wrapFileArray(res.data);
    }
  };
}