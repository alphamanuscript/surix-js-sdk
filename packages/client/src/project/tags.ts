import { AxiosInstance } from 'axios';
import { ProjectTags, Tag, TagList  } from '../types';

export function getProjectTags (projectId: string, apiClient: AxiosInstance): ProjectTags {
  return {
    async list (): Promise<TagList> {
      const res = await apiClient.get<TagList>(`/projects/${projectId}/tags`);
      return res.data;
    }, 
    async rename (oldName: string, newName: string): Promise<Tag> {
      const res = await apiClient.put<Tag>(`/projects/${projectId}/tags/${oldName}`, {
        name: newName
      });
      return res.data;
    }
  };
}