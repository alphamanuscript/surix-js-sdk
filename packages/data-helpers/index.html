<script src="./dist/data-helpers.min.js"></script>