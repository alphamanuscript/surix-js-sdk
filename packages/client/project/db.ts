import { ApiEntity, EntityWrapper } from '@surix/data-helpers';
import { AxiosInstance } from 'axios';
import { ProjectDatabase, Query, TagList } from '../types';

export function getProjectDb (projectId: string, apiClient: AxiosInstance): ProjectDatabase {
  return {
    async get (entityId: string): Promise<EntityWrapper> {
      const res = await apiClient.get<ApiEntity>(`/projects/${projectId}/entities/${entityId}`);
      return new EntityWrapper(res.data);
    },
    async query(query: Query): Promise<EntityWrapper[]> {
      const res = await apiClient.post<ApiEntity[]>(`/projects/${projectId}/entities/query`, query);
      const entities = res.data;
      return entities.map(entity => new EntityWrapper(entity));
    },
    async getTags (): Promise<TagList[]> {
      const res = await apiClient.get<TagList[]>(`/projects/${projectId}/tags`);
      return res.data;
    }
  };
}