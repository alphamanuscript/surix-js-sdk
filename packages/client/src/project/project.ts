import { AxiosInstance } from 'axios';
import { getProjectDb } from './db';
import { getProjectFiles } from './files';

import { Project } from '../types';

export function getProjectApi (projectId: string, apiClient: AxiosInstance): Project {
  return {
    db: getProjectDb(projectId, apiClient),
    files: getProjectFiles(projectId, apiClient)
  };
}