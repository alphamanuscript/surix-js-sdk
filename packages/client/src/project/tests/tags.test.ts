import * as dataHelpers from '@surix/data-helpers';
import { AxiosInstance } from 'axios';
import * as api from '../../api';
import { getProjectApi } from '../project';

describe('Project tags', () => {
  describe('list', () => {
    let apiClient: AxiosInstance;
    const tags = [
      { name: 'posts' },
      { name: 'users' }
    ];
    function callMockedList(): Promise<any[]> {
      apiClient = api.getApiClient('http://baseurl');
      jest.spyOn(apiClient, 'get').mockReturnValue(
        Promise.resolve({ data: tags })
      );
      const project = getProjectApi('project1', apiClient);
      return project.tags.list();
    }
    it('should call projects/:pid/tags endpoint', async () => {
      await callMockedList();
      expect(apiClient.get).toHaveBeenCalledWith(
        '/projects/project1/tags'
      );
    });
    it('should return tags list', async () => {
      const tags = await callMockedList();
      expect(tags).toMatchSnapshot();
     });
  });
});