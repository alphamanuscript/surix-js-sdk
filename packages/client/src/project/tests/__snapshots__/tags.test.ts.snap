// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Project tags list should return tags list 1`] = `
Array [
  Object {
    "name": "posts",
  },
  Object {
    "name": "users",
  },
]
`;
